### PR TITLE
Add cookie support with cmdline switch --cookie

### PR DESCRIPTION
**Pull request by qip**
Cookies are handled by cookiejar in memory with fshttp module through
the entire session.
One useful scenario is, with HTTP storage system where index server
adds authentication cookie while redirecting to CDN for actual files.
Also, it can be helpful to reuse fshttp in other storage systems
requiring cookie.

#### What is the purpose of this change?

To make fshttp module support cookiejar.

#### Was the change discussed in an issue or in the forum before?

No.

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/ncw/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [ ] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/ncw/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [ ] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

**Comment by qip** [comment]
I'm not a golang expert, so this commit has only passed existing tests, anyone would like to spare some time helping me a little?
Anyway, I have a nginx index server which returns 301 to actual files with set-cookie header, and this commit works with that with HTTP backend, so I assume it's functional... Too bad not an actual test case :(

File: fs/config/configflags/configflags.go [hunk 1]
```diff
@@ -87,6 +87,7 @@ func AddFlags(flagSet *pflag.FlagSet) {
 	flags.IntVarP(flagSet, &fs.Config.MaxBacklog, "max-backlog", "", fs.Config.MaxBacklog, "Maximum number of objects in sync or check backlog.")
 	flags.BoolVarP(flagSet, &fs.Config.StatsOneLine, "stats-one-line", "", fs.Config.StatsOneLine, "Make the stats fit on one line.")
 	flags.BoolVarP(flagSet, &fs.Config.Progress, "progress", "P", fs.Config.Progress, "Show progress during transfer.")
+	flags.BoolVarP(flagSet, &fs.Config.Cookie, "cookie", "", fs.Config.Cookie, "Enable session cookiejar.")
```
Comment (ncw) [inline fs/config/configflags/configflags.go]: I think my preference for the name of this flag would be "use-cookies"

Comment (qip) [reply fs/config/configflags/configflags.go]: Actually I'm working on keeping cookies in an external cookiejar file as another option.
Now that rethinking about the naming convention, we could use `--use-cookies` to enable cookie support, and `--cookie-jar-file filename` to load and save cookies (which implies `--use-cookies`); Or, use `--cookies` to enable session cookie, and `--cookies filename` to enable cookie with loading and saving cookies.
How do you think?

**Comment by qip** [comment]
Thank you for reviewing!

> Can you drop all the `docs` changes - I update the docs as part of the release process normally.

Sure, I'll update it together with other updates in next commit.

> Do you think this should be enabled normally?

I think this adds a little bit of overhead, and it doesn't help for backends with proper API like Box or Google Drive, so I'd say we enable it manually by the switch.

**Comment by ncw** [comment]
The CI is complaining about the formatting.  Can you run `fs/fshttp/http.go` through `go fmt` or `goimports`?

**Comment by qip** [comment]
Oh now I see what's the diff in CI log is about. Thanks for the reminder!

**Comment by ncw** [comment]
> Oh now I see what's the diff in CI log is about. Thanks for the reminder!

It is annoying difficult to see! I'm going to fix this shortly by using golangci-lint instead of the rather bodged up system I'm using at the moment!

That looks great now - will merge shortly!

**Pull request merged**